protocols/ping/src/lib.rs

```diff
@@ -226,6 +226,14 @@ where TSocket: AsyncRead + AsyncWrite,
     type Error = IoError;
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.needs_close {
+            match self.inner.close() {
+                Ok(Async::Ready(())) => return Ok(Async::Ready(None)),
+                Ok(Async::NotReady) => (),
```

Comment:
Should this not be `return Ok(Async::NotReady)`? Once we called `close` we must not call `start_send` or `poll_complete` again or else we might get a panic from the inner `Sink` (cf. https://docs.rs/futures/0.1.24/futures/sink/trait.Sink.html#method.close).